File: sdk/python/examples/memory_events_example.py
"""
Memory Events Example

Demonstrates how to listen to memory-related events from the agent:
- Memory stored events
- Memory search events
- Memory recall events
- Memory cleared events
"""

import asyncio
from a3s_code import A3sClient
from a3s_code.types import MemoryItem, MemoryType


async def memory_events_example():
    print("=" * 60)
    print("Memory Events Example")
    print("=" * 60)
    print()

    async with A3sClient(address="localhost:4088") as client:
        # Create a session
        print("1. Creating session...")
        session = await client.create_session(
            name="memory-events-demo",
            workspace="/tmp/memory-events-test",
            system_prompt="You are a helpful coding assistant with memory capabilities.",
        )
        session_id = session["session_id"]
        print(f"✓ Session created: {session_id}")
        print()

        # Track events
        events_received = []

        # Subscribe to events in background
        print("2. Subscribing to events...")

        async def handle_events():
            async for event in client.subscribe_events(session_id=session_id):
                event_type = event.get("type", "")
                events_received.append(event_type)

                if "MEMORY_STORED" in event_type:
                    data = event.get("data", {})
                    print(f"  📝 [MemoryStored] id={data.get('memory_id')}, "
                          f"type={data.get('memory_type')}, "
                          f"importance={data.get('importance')}")

                elif "MEMORIES_SEARCHED" in event_type:
                    data = event.get("data", {})
                    print(f"  🔍 [MemoriesSearched] results={data.get('result_count')}, "
                          f"query={data.get('query')}")

                elif "MEMORY_RECALLED" in event_type:
                    data = event.get("data", {})
                    print(f"  💡 [MemoryRecalled] id={data.get('memory_id')}, "
                          f"relevance={data.get('relevance')}")

                elif "MEMORY_CLEARED" in event_type:
                    data = event.get("data", {})
                    print(f"  🗑️  [MemoryCleared] tier={data.get('tier')}, "
                          f"count={data.get('count')}")

                elif event_type == "EVENT_TYPE_AGENT_END":
                    break

        event_task = asyncio.create_task(handle_events())
        print("✓ Event listener started")
        print()

        # =====================================================================
        # Perform Memory Operations (triggers events)
        # =====================================================================
        print("3. Storing memories (watch for events)...")
        for i in range(3):
            await client.store_memory(
                session_id=session_id,
                memory=MemoryItem(
                    content=f"Test memory {i + 1}: learned about feature #{i + 1}",
                    importance=0.5 + (i * 0.2),
                    tags=["test", f"memory-{i + 1}"],
                    memory_type=MemoryType.EPISODIC,
                ),
            )
            await asyncio.sleep(0.1)  # Small delay to see events
        print()

        print("4. Searching memories (watch for events)...")
        await client.search_memories(
            session_id=session_id,
            tags=["test"],
            limit=10,
        )
        await asyncio.sleep(0.1)
        print()

        print("5. Getting memory statistics...")
        stats_result = await client.get_memory_stats(session_id=session_id)
        stats = stats_result.get("stats")
        if stats:
            print(f"  Long-term: {stats.long_term_count}")
            print(f"  Short-term: {stats.short_term_count}")
            print(f"  Working: {stats.working_count}")
        print()

        print("6. Clearing working memory (watch for events)...")
        await client.clear_memories(
            session_id=session_id,
            clear_long_term=False,
            clear_short_term=False,
            clear_working=True,
        )
        await asyncio.sleep(0.5)
        print()

        # Wait for events to be processed
        try:
            await asyncio.wait_for(event_task, timeout=3.0)
        except asyncio.TimeoutError:
            event_task.cancel()

        # Summary
        print("=" * 40)
        print(f"Event summary:")
        print(f"  Total events received: {len(events_received)}")
        if events_received:
            unique = set(events_received)
            print(f"  Event types: {', '.join(sorted(unique))}")
        print()

        # Clean up
        print("7. Cleaning up...")
        await client.destroy_session(session_id)
        print("✓ Session destroyed")
        print()

        print("=" * 60)
        print("Memory events example complete! ✓")
        print("=" * 60)


if __name__ == "__main__":
    asyncio.run(memory_events_example())

File: sdk/typescript/examples/src/memory-events-example.ts
/**
 * Memory Events Example
 *
 * Demonstrates how to listen to memory-related events from the agent:
 * - Memory stored events
 * - Memory search events
 * - Memory recall events
 * - Memory cleared events
 */

import { A3sClient } from '@a3s-lab/code';

async function memoryEventsExample(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Memory Events Example');
  console.log('='.repeat(60));
  console.log();

  const client = new A3sClient({
    address: process.env.A3S_ADDRESS || 'localhost:4088',
  });

  try {
    // Create a session
    console.log('1. Creating session...');
    const session = await client.createSession({
      name: 'memory-events-demo',
      workspace: '/tmp/memory-events-test',
      systemPrompt: 'You are a helpful coding assistant with memory capabilities.',
    });
    const sessionId = session.sessionId;
    console.log(`✓ Session created: ${sessionId}`);
    console.log();

    // Track events
    const eventsReceived: string[] = [];

    // Subscribe to events in background
    console.log('2. Subscribing to events...');
    const eventPromise = (async () => {
      for await (const event of client.subscribeEvents(sessionId)) {
        const eventType = event.type || '';
        eventsReceived.push(eventType);

        if (eventType.includes('MEMORY_STORED')) {
          console.log(`  📝 [MemoryStored] id=${event.data?.memoryId}, type=${event.data?.memoryType}`);
        } else if (eventType.includes('MEMORIES_SEARCHED')) {
          console.log(`  🔍 [MemoriesSearched] results=${event.data?.resultCount}, query=${event.data?.query}`);
        } else if (eventType.includes('MEMORY_RECALLED')) {
          console.log(`  💡 [MemoryRecalled] id=${event.data?.memoryId}, relevance=${event.data?.relevance}`);
        } else if (eventType.includes('MEMORY_CLEARED')) {
          console.log(`  🗑️  [MemoryCleared] tier=${event.data?.tier}, count=${event.data?.count}`);
        } else if (eventType === 'EVENT_TYPE_AGENT_END') {
          break;
        }
      }
    })();

    console.log('✓ Event listener started');
    console.log();

    // Perform memory operations (triggers events)
    console.log('3. Storing memories (watch for events)...');
    for (let i = 0; i < 3; i++) {
      await client.storeMemory(sessionId, {
        content: `Test memory ${i + 1}: learned about feature #${i + 1}`,
        importance: 0.5 + i * 0.2,
        tags: ['test', `memory-${i + 1}`],
        memoryType: 'MEMORY_TYPE_EPISODIC',
      });
      await new Promise((r) => setTimeout(r, 100));
    }
    console.log();

    console.log('4. Searching memories (watch for events)...');
    await client.searchMemories(sessionId, { tags: ['test'], limit: 10 });
    await new Promise((r) => setTimeout(r, 100));
    console.log();

    console.log('5. Getting memory statistics...');
    const stats = await client.getMemoryStats(sessionId);
    if (stats.stats) {
      console.log(`  Long-term: ${stats.stats.longTermCount}`);
      console.log(`  Short-term: ${stats.stats.shortTermCount}`);
      console.log(`  Working: ${stats.stats.workingCount}`);
    }
    console.log();

    console.log('6. Clearing working memory (watch for events)...');
    await client.clearMemories(sessionId, {
      clearLongTerm: false,
      clearShortTerm: false,
      clearWorking: true,
    });
    await new Promise((r) => setTimeout(r, 500));
    console.log();

    // Wait for events with timeout
    await Promise.race([
      eventPromise,
      new Promise((r) => setTimeout(r, 3000)),
    ]);

    // Summary
    console.log('='.repeat(40));
    console.log('Event summary:');
    console.log(`  Total events received: ${eventsReceived.length}`);
    if (eventsReceived.length > 0) {
      const unique = [...new Set(eventsReceived)];
      console.log(`  Event types: ${unique.sort().join(', ')}`);
    }
    console.log();

    // Clean up
    console.log('7. Cleaning up...');
    await client.destroySession(sessionId);
    console.log('✓ Session destroyed');
    console.log();

    console.log('='.repeat(60));
    console.log('Memory events example complete! ✓');
    console.log('='.repeat(60));
  } catch (error) {
    console.error('Error:', error);
  }
}

memoryEventsExample();
